enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,1051 @@
+- Compare versions (metadata vs binary)
+  - Same - start cluster
+  - Different - means that both binaries implement the enhancement,
+    so this could be considered a bug.
+  - Metadata is also missing
+    - Assume data was created by 4.13, do a backup and follow data migration rules
+
+#### When should version metadata file be created
+- By the green script
+  - Existence of file would indicate that data belongs to healthy deployment
+    - What upon reboot? Should file be cleared to be later recreated by green script?
+      - Blind assumption `.version exists == healthy data` might be too implicit and lead
+        to wrong conclusions
+- On start of `microshift run` or shortly after with addition from green and red scripts
+  - `{ "version": "X.Y.Z", "status": "unknown"}` would be created at first
+  - Green script would change `.status` to `healthy`
+  - Red script would change `.status` to `unhealthy`
+  - Pros:
+    - Explicit
+  - Cons:
+    - More complicated schema - weak con
+
+#### How should `microshift pre-run` be executed?
+- `microshift.service` - `ExecStartPre`
+  - No need to add new systemd service files.
+  - It will run on each `systemctl restart microshift` which is not desirable (will it run when systemd restarts MicroShift?)
+- `microshift-pre-run.service`
+  - Running on boot, just once, before `microshift.service`
+  - Not repeated on MicroShift restart
+  - New service file
+
+#### Should MicroShift healthcheck check and log version skew problems so it's easier to debug?
+- Why not?
+
+#### Should backups be kept only for deployments present in ostree command?
+- Is it possible that deployment can be reintroduced?
+  - I.e. it will have the same id and admin might want to rollback to it?
+
+#### Data exists, action is "restore", but backup does not exists
+- It's different from scenario where backup exists, but it fails to be restored
+- See flows below for more context:
+  - ["Rollback to first deployment, failed restore"](#rollback-to-first-deployment-failed-restore)
+  - ["Fail first startup, FDO (FIDO Device Onboard) deployment"](#fail-first-startup-fdo-fido-device-onboard-deployment)
+- Note: this logic is part of `microshift pre-run` which exit code will affect if 
+  `microshift run` executes or not
+- Ideas so far:
+  - Remove the data and allow for `microshift run`
+    - Potential data loss?
+      - MicroShift data is not a biggest problem, it would be more of an inconvenience
+        due to cert regeneration, but static manifests will be applied again
+      - Bigger impact: customer's app loosing reference to PVC it was using
+        - This is also loaded scenario, because if we restore, everything should restore
+          so app shouldn't rely on "future" (from its version perspective) data
+        - Let's assume that thin-snapshot is made, and the same PVC goes across ostree
+          deploys. Rollback/restore would restore thin-snapshot, so app should be
```

Comment:
Does changing the ostree deployment affect the LVM system? I would expect it to be limited to the filesystem.

Reply:
Not the lvm itself, but connection between topolvm (which data resides in etcd) and lvm.
If we cleanup the microshift data, topolvm data is also lost so volume that was assigned to the app via PVC is now dangling and app will get new, clean volume.

Reply:
Backup of customer application bleeding in

Reply:
Yeah, it might be clearer to leave out any mention of app data. We say that is out of scope for this enhancement.